### PR TITLE
Source plan environment variables after `prepare` and `execute` steps

### PR DESCRIPTION
**Pull request by therazix**
Environment variables can now be written inside the `TMT_PLAN_ENVIRONMENT_FILE`. Variables defined in this file are sourced immediately after the `prepare` step, making them accessible in the tests and across all subsequent steps. Variables inside the `TMT_PLAN_ENVIRONMENT_FILE` have a lower priority than `environment` and `environment-file` fmf keys and can be overwritten by them.

Pull Request Checklist:

* [x] implement the feature
* [x] write the documentation
* [x] extend the test coverage
* [x] mention the version

Fixes: #2276

### CODE REVIEW

File: tmt/base.py [hunk 1]
```diff
@@ -1722,6 +1735,20 @@ def _initialize_data_directory(self) -> None:
...
+
+        Environment variables stored in this file are sourced
+        before the execute step and are available to all
+        subsequent steps.
+        """
+        assert self.data_directory is not None  # narrow type
+        self.env_file = self.data_directory / "variables.env"
+        self.debug(
+            f"Create the environment file '{self.env_file}'.", level=2)
+        self.env_file.touch(exist_ok=True)
```
Comment (happz) [inline tmt/base.py:1750]: This would be a nice cached property - the first time it's accessed, the file would be created.

```python
@tmt.utils.cached_property
def env_file(self) -> Path:
    assert self.data_directory is not None  # narrow type

    env_file_path self.data_directory / "variables.env"
    env_file_path.touch(exist_ok=True)

    self.debug(f"Create the environment file '{env_file_path}'.", level=2)

    return env_file_path
```

Path doesn't change, it's a property, no need to call it explicitly when we can forget about it, and it'll materialize the first time tmt tries to add it to a step's environment.

one think I'm worried about though is the name - there's already `environment` and `environment-file`, not `env_file`. It would be nice to tell the difference between `environment-file` and `env_file`, but I have no idea what to propose :/

File: tmt/base.py [hunk 2]
```diff
@@ -2187,6 +2214,9 @@ def go(self) -> None:
                         color='yellow', shift=1)
                     abort = True
                     return
+                # Source the plan environment file after prepare step
+                if isinstance(step, tmt.steps.prepare.Prepare):
+                    self._source_plan_env_file()
```
Comment (happz) [inline tmt/base.py:2219]: I suppose `execute` step may add variables to the file as well, hoping to expose them to `finish` phases. Re-including the file once more after `execute` would allow that.

File: tmt/base.py (hunk 1 above)
Comment (therazix) [reply tmt/base.py]: Thank you for the suggestion. I added the cached property.

I agree that the name is quite confusing, but I'm also not sure what would be the best name for it.

File: tmt/base.py (hunk 2 above)
Comment (therazix) [reply tmt/base.py:2216]: I changed it so the file is sourced after the execute step as well. However, I encountered a problem with the `environment_file_to_dict` function I used because it was cached. So, I used the `__wrapped__` attribute to ignore the cache. Is there a better way to do that, or is this approach okay?

Comment (happz) [reply tmt/base.py:2216]: `__wrapped__` seems to be fine. There's `clear_cache()` but that seems to be a hammer way too big.

File: tmt/base.py [hunk 3]
```diff
@@ -1553,6 +1553,7 @@ def __init__(
                 self._initialize_worktree()
 
             self._initialize_data_directory()
+            self._plan_environment: EnvironmentType = {}
```
Comment (happz) [inline tmt/base.py:1556]: I'd recommend moving somewhere on line 1566, after `self._environment` initialization: the attribute may exist all the time, it'd just remain empty, making the reasoning about it easier.

The name, the name again! :( I still don't have a better name, but I'd like to see some uniformity: there's `_plan_environment`, `env_file`, and `_source_plan_env_file`. How about renaming the latter two to follow the example of the first one, and stick to "plan_environment": `_plan_environment` to hold the loaded envvars, `plan_environment_file` property providing the path, and `_source_plan_environment_file` populating `_plan_environment` from the said file. When one starts looking for "plan environment", the `plan_environment` substring would point to all bits taking care of its processing.

Comment (therazix) [reply tmt/base.py]: Fixed in e70b67e.

**Review by happz — APPROVED** [review]

File: docs/overview.rst [hunk 4]
```diff
@@ -430,6 +430,11 @@ TMT_PLAN_DATA
     back from the guest and available for inspection after the
     plan is completed.
 
+TMT_PLAN_ENVIRONMENT_FILE
+    Path to the file containing environment variables that should
+    be sourced after prepare and execute steps. These variables will
+    then be accessible for all subsequent steps.
```
Comment (psss) [inline docs/overview.rst:436]: Perhaps we should explicitly mention the expected format of the file? And include an example to make it perfectly clear?

File: tests/core/plan-env-file/data/plan.fmf [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    /env-file-override:
+        prepare:
+            how: shell
+            script: echo "MYVAR1=MYVAR1_VALUE" >> $TMT_PLAN_ENVIRONMENT_FILE
+        environment-file:
+            - override.env
+
+    /execute-override:
+        discover:
+            how: shell
```
Comment (psss) [inline tests/core/plan-env-file/data/plan.fmf:34]: The `discover` step here is not needed as `script` is used under `execute`. Or we can move the test script here.

File: tests/core/plan-env-file/data/plan.fmf [hunk 6]
```diff
@@ -0,0 +1,40 @@
+/test:
```
Comment (psss) [inline tests/core/plan-env-file/data/plan.fmf:1]: Let's call the file `main.fmf` as `plan` becomes part of the `fmf` name:

```
Found 1 test: /plan/test.
Found 4 plans: /plan/plan/env-file-override, /plan/plan/execute-override, /plan/plan/multiple and /plan/plan/single.
```

File: tests/core/plan-env-file/test.sh [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+. /usr/share/beakerlib/beakerlib.sh || exit 1
+
+rlJournalStart
+    rlPhaseStartSetup
+        rlRun 'pushd data'
+        rlRun 'set -o pipefail'
+    rlPhaseEnd
+
+    rlPhaseStartTest 'Check environment variable is available'
+        rlRun -s 'tmt run -vvv plan --name /plan/single'
```
Comment (psss) [inline tests/core/plan-env-file/test.sh:11]:
```suggestion
        rlRun -s 'tmt run -rvvv plan --name /plan/single'
```

Let's clean up after ourselves so that many run dirs are not left behind when running tests locally.

File: tmt/base.py [hunk 8]
```diff
@@ -1682,6 +1687,16 @@ def environment(self) -> EnvironmentType:
             return combined
         return self._environment
 
+    def _source_plan_environment_file(self) -> None:
+        """ Add variables from the plan environment file to the environment """
+        if (self.my_run and self.plan_environment_file.exists()
+                and self.plan_environment_file.stat().st_size > 0):
+            # using __wrapped__ to force the reload of the file
```
Comment (psss) [inline tmt/base.py:1694]:
```suggestion
            # Use __wrapped__ to force the reload of the file
```

Just a consistency nitpick.

File: docs/overview.rst (hunk 4 above)
Comment (psss) [reply docs/overview.rst:436]: For some recent variables we've also included a `versionadded` mention. @lukaszachy, @happz, would it make sense here as well?

**Review by psss — CHANGES_REQUESTED** [review]
Thanks a lot for implementing this! Looks very good. Added just a few minor suggestions.

File: docs/overview.rst (hunk 4 above)
Comment (therazix) [reply docs/overview.rst]: Added in 9ecf336.

File: tests/core/plan-env-file/data/plan.fmf (hunk 5 above)
Comment (therazix) [reply tests/core/plan-env-file/data/plan.fmf]: Fixed in 9ecf336.

File: tests/core/plan-env-file/data/plan.fmf (hunk 6 above)
Comment (therazix) [reply tests/core/plan-env-file/data/plan.fmf]: Fixed in 9ecf336.

File: tests/core/plan-env-file/test.sh (hunk 7 above)
Comment (therazix) [reply tests/core/plan-env-file/test.sh]: Fixed in 9ecf336.

File: tmt/base.py (hunk 8 above)
Comment (therazix) [reply tmt/base.py]: Fixed in 9ecf336.

File: docs/overview.rst [hunk 9]
```diff
@@ -430,6 +430,21 @@ TMT_PLAN_DATA
...
+    the file has to be in the format of ``NAME=VALUE`` and each
+    variable should be on a separate line.
+
+    Example of the file content::
+
+        COUNT=1
+        VARIABLE=VALUE
+        ANOTHER_VARIABLE=ANOTHER_VALUE
+
+.. versionadded:: 1.29
```
Comment (happz) [inline docs/overview.rst:446]: I think this should be indented to the level of the text above, to be rendered for the text above.

**Review by happz — APPROVED** [review]

File: docs/overview.rst (hunk 9 above)
Comment (therazix) [reply docs/overview.rst]: Fixed in 8ec6eee.

**Review by psss — APPROVED** [review]
Now looks good. I just removed an extra `script` key, fixed typo and indented the other `versionadded` in the variables overview in c0599e45.

**Comment by psss** [comment]
One question though:

    Variables inside the `TMT_PLAN_ENVIRONMENT_FILE` have a lower
    priority than `environment` and `environment-file` fmf keys
    and can be overwritten by them.

Wouldn't it be better to switch the priority to allow overriding environment defined in the plan from the running process? I can imagine to put some default value in the plan config and then, based on a specific situation/condition, modify it during `prepare` or `execute` step. @happz, @lukaszachy, thoughts?

**Review by psss — CHANGES_REQUESTED** [review]
Summary from the hacking session: The new variables should have a higher priority and override any environment variables defined in the plan. @therazix, could you please update the pull request accordingly?

**Comment by psss** [comment]
For reference, summary from the adhoc chat:

* Revert the change to make plan environment stronger than `TMT_PLAN_ENVIRONMENT_FILE`
* Explicitly document the priority in the variable definition to make it clear

Reasoning:

* This is kind of consistent with the fact that plan environment is stronger than test environment
* Using this approach we will avoid documenting weird corner case for the `script` field in plan
* The other approach would require to somehow ignore certain fields in `Plan._expand_node_data()`

**Comment by therazix** [comment]
The change was reverted. The variables specified by the `TMT_PLAN_ENVIRONMENT_FILE` now have lower priority than variables specified by the `environment` key, `environment-file` key, or the command line. I also documented the priority and added some more tests.

**Review by psss — APPROVED** [review]
Looks good, thanks!
